### ISSUE TITLE
[Python] Implement new Concept Exercise: dicts

### ISSUE THREAD

**Issue by BethanyG** [exercism/v3#877]
This issue describes how to implement the `dicts` concept exercise for the Python track.  

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the basics of the `dict` (dictionary, mapping) data type in Python. 

## Learning objectives

- create a `dict` via constructor and literal
- access values in a `dict` via `keys` 
- assign values in a `dict` via key reference 
- check for membership of a key in a given dictionary
- add new `key`:`value` pairs to the `dict`
- remove `key`:`value` pairs from the `dict`.
- iterate through a `dict` using `dict.keys()`, `dict.values()`, or `dict.items()`
- `dict` method `setdefault()`

## Out of scope

- Dictionary comprehensions
- `dict` methods such as `get()` or `clear()`
- Built-in functions as they relate to this data structure (**e.g.** `copy()`, `len()`, or `enumerate()`
- Sorting by `keys` or `values` 
- Swapping `keys` for `values` 
- Knowing that Dictionaries can be _**nested**_, -- **e.g.** " a dictionary of dictionaries".
- Mutability
- `copy() vs `deepcopy()`
-  Related `collections` module with `Counter()` and `defaultdict()`
-  Memory and performance characteristics.

## Concepts

- dicts

## Prerequisites

- [builtin-types](https://github.com/exercism/v3/tree/master/languages/python/reference/concepts/builtin_types)
- [data-types](https://github.com/exercism/v3/blob/master/languages/python/reference/concepts/data_structures.md)

## Resources to refer to

- [Dictionaries (Python official docs)](https://docs.python.org/3/tutorial/datastructures.html#dictionaries)
- [Mapping Types (Python official docs)](https://docs.python.org/3/library/stdtypes.html#mapping-types-dict)

### Hints

Hints should link to the `Dictionaries` section of the Python docs tutorial:  [Dictionaries](https://docs.python.org/3/tutorial/datastructures.html#dictionaries).

### After

After, the student can explore comprehension syntax, although it will be taught in separate exercises.  This would also be a good time to explore how built-in methods such as `sorted()` and  `len()` can be used with dictionaries, and how default values (through `setdefault()`) can be used as a `dict` is being created.

## Representer

No changes required.

## Analyzer

No changes required.

## Implementing

Tests should be written using unittest.TestCase, and the test file named dicts_basic_test.py.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

## Edits
- edited for updated naming by @yawpitch
- edited to amend things to learn by @BethanyG

**Comment by J08K** [comment]
### **Hey there**
My name is Job and I'd love to be assigned to this concept exercise. I already put a short, not-fleshed out version of this concept on the #maintaining-python channel on Slack.

### **_Here is my idea:_**
The exercise is called _Inventory management_ it will teach the student how to add items, remove items, iterate through a dictionary. 
### **_What does the exercise look like?_**
The student will get two lists (products and codes), the products should be put in inventory according to their corresponding codes (products and codes are ordered in the same way). The user has to put the items in a dictionary as follows: `{product : code}`.
### _**The catch:**_
Some codes are invalid, the code should look like this: `XXX-XXA-XXX`. _A_ is a control number, that number should always be between 4 and 10 [5-9]. It should also not be any longer or shorter than 11 characters. If the code is invalid, it should not be included in the dictionary, neither should its corresponding product. It can simply be deleted.
### _**Here is what the code could look like:**_
_This is a sample solution:_
```
def manage(products, codes):
    inventory = dict()
    for product, code in zip(products, codes):
        inventory[product] = code
    for product in inventory:
        if int(inventory[product][6]) < 5:
            del inventory[product]
        if len(inventory[product]) > 11 or len(inventory[product]) < 11:
            del inventory[product]
    return inventory
```
The code could probably be improved further, but this is just a proof of concept anyway. I might also do tests with letters in the codes, which this solution is not capable of handling, yet.
```
import unittest
from inventory_management import manage

class inventory_management_test(unittest.TestCase):
    def all_codes_correct(self):
        self.assertEqual(manage(["Ducks", "Geese", "Swan"], ["111-226-895", "123-456-789", "158-765-589"]),
                                {"Ducks":"111-226-895", "Geese":"123-456-789", "Swan":"158-765-589"})
    def one_code_correct(self):
        self.assertEqual(manage(["Chicks", "Miniducks"], ["243-198-657", "152-420-781"]), {"Chicks":"243-198-657"})
```
I have never used unittest before, so it might look a little messy, but I am going to teach myself how to use it. And yes, there will be more tests added, this is just a sneak-peek. :)

This will, I think, give the student a good view of what dictionaries are and how to use them.

Thank you for reading,
**- Job van der Wal (J08K)**

**Comment by yawpitch** [comment]
Thanks for this. An inventory management story is interesting, however I've got some concerns about this as described. Or perhaps not so much concerns as details to hammer out and address before you start working on the actual code and copy.

1. What are the prerequisite concepts of this concept?  From the example code it would appear the student needs to have been introduced to the `zip` builtin, list indexing, comparison operations on integers, the `del` keyword, and the `len` builtin (which is one of the things in the "Out of Scope" section above), as well as conditionals and loops at a minimum. It makes sense that `lists` would be a pre-requisite concept (assuming the `lists` exercise covers indexing and length), since it's a less-complex builtin data type, but I'm unsure if `zip` needs to have been introduced at this point.
2. Does string validation really need to be part of this exercise? That implicitly adds at least `strings` to the prerequisites, and although that's sensible in terms of ordering and a student should understand strings well enough _before_ looking at dictionaries, does this simple validation actually add anything to understanding dictionaries? Perhaps we should have them actually count items into and out of inventory instead? Like write an "add_item" function that takes a an inventory and an item and adds 1 to the item's count if it exists, or sets the items count to 0 if it does not. Similarly a "remove_item" function could take the same arguments and decrement the count or remove the item entirely once inventory reaches zero. Since nested dicts are outside the scope of this I'd say tracking say price and barcode and inventory count might be left to later exercises.

Again I think the overall idea is great, I just want to ensure we're focusing each concept exercise _very_ tightly on the concept being taught, and that the story doesn't compel us into adding to the student's cognitive workload.

**Comment by J08K** [comment]
Hey @yawpitch , thank you for your feedback! Exactly the type of stuff I needed.

You are definitely right about the amount and difficulty of the prerequisites I chose to use in my concept solution. I completely missed the `len` builtin in the out of scope part, I'm sorry. Although I do wonder why it's in the out of scope part, as I use it quite often myself and it's very handy to know and not be explained after the student has completed the exercise. When I look at the [list's issue](https://github.com/exercism/v3/issues/859) I can see that the `len` builtin will be explained, so why not use it in here? Also, you don't have to use `zip` to make this solution work, but I just used it as an example. Zip isn't important for now, so we can just ignore it in this exercise.

Great idea about the counting of items, that's probably a much better way of explaining dictionaries. So do you want to have two functions? One called "add_item" and one called "remove_item"? I think that's a great way of explaining dictionaries, as it explains adding items and accessing items. I assume that if an item goes below one, it has to be set to zero, right? And not delete the item?

I'll work on this idea and I'll comment on this tonight with some _concept solutions_ and a _concept test file_. If I get the green light, then I will start working on this fully.

Thank you for your feedback! Better discuss this more, than have bad exercises.

- Job van der Wal (J08K)

**Comment by J08K** [comment]
### Concept exercise iteration 2
From the feedback, I got from user @yawpitch I have rethought the new exercise. Mine was a bit too advanced for beginning students. This new concept exercise should be much more applicable to most students beginning with the Python programming language.

### _What is the new idea?_
The exercise will be called _Inventory_. The goal is to create a dictionary with products and the amount of each product. The student will get a list of products, and whenever a given product occurs, the solution will have to add 1 of that product to the dictionary. The dictionary will look like this: `{product : amount}`. There will also be a second function that removes 1 amount of a given product from the list, this will go down to 0 and not further (item won't get deleted from the dictionary).

### _So?_
The first function, called "add_items" will add the items in the given list. This list will look like the following: `["Duck", "Swan", "Goose", "Duck"]`. When given this list, the solution should output a dictionary that looks like this: `{"duck":2, "swan":1, "goose":1}. As you can see, all the letters should be lowered, this is because _products_ spelled like "GoOsE" should be the same as "goose".

The second function, called "remove_items". This function will receive a list and a dictionary, the list contains the products that should have 1 removed from their key in the dictionary. **Although I am not sure _if_ we should test what happens when a product has been given that isn't in the given dictionary. Raise a _ValueError_? Just ignore that product?** So if the input is `(["Duck", "Goose"], {"Duck":2, "Goose":1, "Swan":1})` the output should be: `{"duck":1, "goose":0, "swan":1}`.

### _What would the code look like?_
This is an example "add_items" function for the solution of the exercise:
```
def add_items(products):
    dictionary = dict()
    for item in products:
        if item.lower() not in dictionary:
            dictionary[item.lower()] = 1
        else:
            dictionary[item.lower()] += 1
    return dictionary

```
This is an example "remove_items" function for the solution of the exercise:
```
def remove_items(to_remove, products):
    for item in to_remove:
        if item.lower() in products:
            if products[item.lower()] != 0:
                products[item.lower()] -= 1
    return products
```
As I am not entirely familiar with _unittools_ yet, I haven't written a test for this _concept exercise_. I plan on getting familiar with it very soon, as it will also become very handy in the future. So don;t worry.

Thank you for reading, and I hope it's definitely an improvement from the last iteration! If there is something that can still be changed, let me know!

- Job van der Wal (J08K)

PS; If some parts are unclear to you, I am happy to explain.

**Comment by yawpitch** [comment]
@J08K this is indeed simpler, it's basically treating a dict as a generalized _Bag_ data collection. However I'd make a few suggestions.

First, both functions should have the same signature ... they should take a **dict** (the current inventory state) as the first argument and a **list** of **str** (the items to add / remove) as the second. This will allow us to create whatever inventory and items values we wish and hand them in and make it easy for us to predict what the expected value to test for will be.

In the case of `add_items` if the item does not exist in the dict then it should be added with a default count of 0, then incremented. An example implementation would be:

```python
def add_items(inventory, items):
    for item in items:
        inventory.setdefault(item, 0)
        inventory[item] += 1
```
   
We need not have a return value, since this is making an in-place change to the inventory.

In the case of `remove_items` if the item does not exist we should also add it with a default count of 0. If it does exists we should decrement the count, but never below 0. As this basically is a Bag structure you're correct we should not delete the item from inventory, merely reduce its count to at least 0.

```python
def remove_items(inventory, items):
    for item in items:
        inventory.setdefault(item, 0)
        inventory[item] = max(inventory[item] - 1, 0)
```

This involves a number of basic dict operations -- at least **dict.setdefault**, getting and setting items, possibly with **dict.get** -- so it's a small but sensible introduction.

I'd also suggest we have at least one function that involves iterating over a **dict** itself, likely over **dict.items** ... my thought would be something like:

```python
def list_inventory(inventory):
    in_store = [(k, v) for k, v in sorted(inventory.items()) if v]
    return in_store
```

This would add at least the notion of iterating over keys/values/items, and the student should already have sufficient knowledge of lists and tuples by this point to complete it.

Testing for all three should be quite straightforward. I can help you with unittest, if needed.

For simplicities sake we could do the students a favor and guarantee that items will always be lowercase (or upper case), to avoid unnecessary distraction with string methods, and that the inventory will only include items with a count of 0 or more (ie negative counts don't need to be handled). Basically I'd like us to describe each problem very precisely so that there's no tendency for the student to drift off into thinking too much about policing inputs, raising exceptions, etc.

**Comment by J08K** [comment]
Hey @yawpitch 
Sorry that I couldn't reply sooner, I had some things at home that I needed to deal with. I think that we are ready to actually go work on it.

Just so I am sure I understand: You want to have 3 functions called `add_items`, `remove_items`, and `list_inventory`, where the first two don't return a value, but the last one just returns a list of items in tuples, right?

For the rest, yeah we are focussing on the right things now. Looks good.

I will go work on it completely now. This is my first concept exercise, so I have to create the 4 or so files which are used in the concept exercise, right?

Thanks
- Job van der Wal

**Comment by yawpitch** [comment]
@J08K no worries, we've all got lots of home things these days, to say the least. As for the functions, the specifics on return values are up to you; typically functions that intentionally mutate in place an incoming argument _do not_ return a value, which saves confusion for cases like this:

```python
old_inventory = {"socks": 2}
new_inventory = add_items(old_inventory, ["socks", "socks"])
```

From looking at that it's not clear if `assert old_inventory is new_inventory` should fail or succeed. So better to return None and indicate in the docstring that the function intentionally mutates.

Also it's up to you if you want to make the **list_inventory** return a tuple of items or a formatted string ... I'm not even dead set on that function, I just think that there should be _some_ function that demonstrates iterating over a dictionary. Ideally we'd introduce **dict.keys**, **dict.values** and **dict.items** individually, but I thought listing the inventory would be one way of addressing at least one of them. Feel free to come up with others.

And yes you do have to create the files, however I've written a little utility that will build out the basic skeleton for you, see the README in https://github.com/exercism/v3/tree/master/languages/python/bin for details.

**Comment by BethanyG** [comment]
@J08K - If you haven't already, I would suggest writing up what you have in a `draft PR`.  That way, we could merge when this gets close.  😄   _So happy you are working on this!_
